src/DistributedFFT.chpl

```diff
@@ -212,195 +215,156 @@ prototype module DistributedFFT {
-        const yDest = myDestDom.dim(1);
-        const xDest = myDestDom.dim(2);
-
-        const numTasks = min(here.maxTaskPar, zSrc.size);
-        const numTransforms = zSrc.size/numTasks;
-
-        // Set up FFTW plans
-        var plan_y = setupPlanColumns(T, ftType, {ySrc, zSrc}, numTransforms, signOrKind, FFTW_MEASURE);
-        var plan_y1 = setupPlanColumns(T, ftType, {ySrc, zSrc}, numTransforms+1, signOrKind, FFTW_MEASURE);
-        var plan_x = setupPlanColumns(T, ftType, {xDest, zSrc}, numTransforms, signOrKind, FFTW_MEASURE);
-        var plan_x1 = setupPlanColumns(T, ftType, {xDest, zSrc}, numTransforms+1, signOrKind, FFTW_MEASURE);
-        var plan_z = setup1DPlan(T, ftType, zSrc.size, 1, signOrKind, FFTW_MEASURE);
-
-        // Use this as a temporary work array to
-        // avoid rewriting the src array
-        var myplane : [{0..0, ySrc, zSrc}] T;
-        const z0 = zSrc.first;
-        forall iy in ySrc {
-          c_memcpy(c_ptrTo(myplane[0,iy,z0]),
-                   c_ptrTo(src.localAccess[xSrc.first,iy,z0]),
-                   myLineSize);
+      for ix in xSrc {
+        // Y-transform
+        forall myzRange in batchedRange(zSrc, numTasks) {
+          ref elt = myplane[0, ySrc.first, myzRange.first];
+          if myzRange.size == batchSizeSm then yPlanSm.execute(elt);
+                                          else yPlanLg.execute(elt);
         }
 
-        for ix in xSrc {
-          // y-transform
-          const y0 = ySrc.first;
-          forall myzRange in batchedRange(zSrc) with (ref plan_y, ref plan_y1) {
-            select myzRange.size {
-                when numTransforms do plan_y.execute(myplane[0,y0,myzRange.first]);
-                when numTransforms+1 do plan_y1.execute(myplane[0,y0,myzRange.first]);
-                otherwise halt("Bad myzRange size");
-              }
-          }
-
-          // z-transform
-          // Offset to reduce collisions
-          const offset = (ySrc.size/numLocales)*here.id;
-          forall iy in 0.. #ySrc.size with (ref plan_z) {
-            const iy1 = (iy + offset)%ySrc.size + y0;
-            plan_z.execute(myplane[0,iy1,z0]);
-            // This is the transpose step
-            remotePut(dest[iy1,ix,z0], myplane[0,iy1,z0], myLineSize);
-            // If not last slice, copy over
-            if (ix != xSrc.last) {
-              c_memcpy(c_ptrTo(myplane[0,iy1,z0]),
-                       c_ptrTo(src.localAccess[ix+1,iy1,z0]),
-                       myLineSize);
-            }
+        // Z-transform, offset to reduce comm congestion/collision
+        forall iy in offset(ySrc) {
+          zPlan.execute(myplane[0, iy, zSrc.first]);
+          // This is the transpose step
+          copy(Dst[iy, ix, zSrc.first], myplane[0, iy, zSrc.first], myLineSize);
+          // If not last slice, copy over
+          if (ix != xSrc.last) {
+            copy(myplane[0, iy, zSrc.first], Src[ix+1, iy, zSrc.first], myLineSize);
```

Comment:
+1 for the uniformity. Are we sure we're not losing on performance here by not dropping to a `memcpy` in the fully local case?

Reply:
Ignore this -- I missed the piece in your PR that mentioned that there was no loss in performance!

Reply:
Yeah, no perf loss. It ends up turning into a chpl_comm_get, that short-circuits to a memmove. I think theres like 2 extra branches and a function call, but in the context of a large memcpy that's negligible. 

And from previous work I did, memcpy and memmove have identical performance for larger transfers (aside, but intel doesn't optimize small memmove like they do memcpys -- https://github.com/chapel-lang/chapel/pull/11473#issuecomment-435992457)